internal/writeaheadlog/wal.go

```diff
@@ -0,0 +1,295 @@
+
+// rotate a log file
+// it finalizes an active file if it is open
+// and opens a new one
+func (wal *WriteAheadLog[T, PT]) rotate() error {
+	if wal.active.file != nil {
+		if err := wal.finalize(); err != nil {
+			return fmt.Errorf("finalizing log file: %w", err)
+		}
+	}
+
+	wal.active.logname = time.Now().UTC().Format(time.RFC3339Nano) + walExtension
+	var err error
+	wal.active.file, err = os.OpenFile(
+		filepath.Join(wal.path, wal.active.logname),
+		os.O_CREATE|os.O_WRONLY|os.O_EXCL, 0666)
+
+	if err != nil {
+		wal.active.file = nil
+		return fmt.Errorf("opening new log file %q: %w", wal.active.logname, err)
+	}
+	wal.active.cborWriter = cbg.NewCborWriter(wal.active.file)
+	return nil
+}
+
+// finalize the existing log
+func (wal *WriteAheadLog[T, PT]) finalize() error {
+	if wal.active.file == nil {
+		return nil
+	}
+	err := wal.active.file.Sync()
+	if err != nil {
+		return fmt.Errorf("syncing content to disk: %w", err)
+	}
+	err = wal.active.file.Close()
+	if err != nil {
+		return fmt.Errorf("closing the file: %w", err)
+	}
+	wal.active.file = nil
+	wal.active.cborWriter = nil
+
+	wal.active.logContent.complete = true
+	wal.logFiles = append(wal.logFiles, wal.active.logContent)
+	wal.active.logContent = logContent[T, PT]{}
+	return nil
+}
+
+func (wal *WriteAheadLog[T, PT]) readInAll() error {
+	_, err := os.Stat(wal.path)
+	if errors.Is(err, os.ErrNotExist) {
+		err := os.MkdirAll(wal.path, 0777)
+		if err != nil {
+			return fmt.Errorf("making WAL directory at %q: %w", wal.path, err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("checking WAL dir: %w", err)
+	}
+
+	dirEntries, err := os.ReadDir(wal.path)
+	if err != nil {
+		return fmt.Errorf("reading dir entries at %q: %w", wal.path, err)
+	}
+	slices.SortFunc(dirEntries, func(a, b fs.DirEntry) int {
+		return strings.Compare(a.Name(), b.Name())
+	})
+
+	for _, entry := range dirEntries {
+		if !strings.HasSuffix(entry.Name(), walExtension) {
+			continue
+		}
+		content, err := wal.readLogFile(entry.Name())
+		if err != nil {
+			return fmt.Errorf("readling log file: %w", err)
+		}
+
+		wal.logFiles = append(wal.logFiles, content)
+	}
+	return nil
+}
+
+// readLogFile reads in a file logname as a log
+// it eats parsing errors and logs in on purpose
+func (wal *WriteAheadLog[T, PT]) readLogFile(logname string) (logContent[T, PT], error) {
+	var res logContent[T, PT]
+	logFile, err := os.Open(filepath.Join(wal.path, logname))
+	if err != nil {
+		return res, fmt.Errorf("opening logfile %q: %w", logname, err)
+	}
+	defer logFile.Close()
+
+	logFileReader := cbg.NewCborReader(logFile)
+	var content []T
+	var maxEpoch uint64
+	for err == nil {
+		var single T
+		err = PT(&single).UnmarshalCBOR(logFileReader)
+		if err == nil {
+			content = append(content, single)
+			maxEpoch = max(maxEpoch, PT(&single).WALEpoch())
+		}
+	}
+
+	res = logContent[T, PT]{
+		logname:  logname,
+		content:  content,
+		maxEpoch: maxEpoch,
+		complete: true,
+	}
+
+	switch {
+	case err == nil:
```

Comment:
this case isn't even possible, right?

Reply:
I don't think it is, I expect us to hit EOF or UnexpectedEOF.